src/libclipper/include/clipper/threadpool.hpp

```diff
@@ -194,6 +194,7 @@ class ThreadPool {
     boost::shared_lock<boost::shared_mutex> l(queues_mutex_);
     auto queue = queues_.find(queue_id);
     if (queue != queues_.end()) {
+      l.unlock();
```

Comment:
Why do you unlock here? It looks like this would be unsafe,  and the lock will be released very soon anyway (as soon as the function completes and `l` goes out of scope).

Reply:
@dcrankshaw : 
I think that `queues_mutex_` is a lock to resolve mutual exclusion condition for `queues_` only, not for `ThreadSafeQueue` (the same with queue->second).

Reply:
Agreed. But `queue` is an iterator that points to an entry in the `queues_` object. If `queues_` is modified, the iterator can become invalid and the `queue->second.push(...)` call could result in a segfault. Also, note that the lock `l` is a shared lock, so in the common case when you have multiple threads adding work to queues, those threads can all acquire the lock and won't block each other.

Reply:
I understand. In line with your comment, I think that [worker codes](https://github.com/ucbrise/clipper/pull/519/files#diff-0ab69d323c13a62e04fbe37aa268462bR212) must be reverted also. How we can resolve this problem?